### PR TITLE
an opinionated action modal

### PR DESCRIPTION
**Pull request by UnsungHero97**
Examples:

1. basic
```
<ActionModal
    isVisible={isVisible}
    onClickPrimary={onClickPrimary}
    onClickSecondary={closeModal}
    onClose={closeModal}
    requestState={requestState}
    textTitle="Confirm Action" />
```

2. custom modal body based on request state
```
const requestStateComponents = {
  [RequestStates.STANDBY]: (
    <span>Please confirm you want to take this action.</span>
  ),
  [RequestStates.SUCCESS]: (
    <span>The action succeeded.</span>
  ),
  [RequestStates.FAILURE]: (
    <span>The action failed. Please try again.</span>
  ),
};

return (
  <ActionModal
    isVisible={isVisible}
    onClickPrimary={onClickPrimary}
    onClickSecondary={closeModal}
    onClose={closeModal}
    requestState={requestState}
    requestStateComponents={requestStateComponents}
    textPrimary="Yes, confirm"
    textSecondary="No, cancel"
    textTitle="Confirm Action" />
);
```

STANDBY
Screen Shot 2019-10-28 at 5 57 08 PM

PENDING
Screen Shot 2019-10-28 at 5 58 16 PM

SUCCESS
Screen Shot 2019-10-28 at 5 58 24 PM

FAILURE
Screen Shot 2019-10-28 at 5 58 43 PM

### CODE REVIEW

File: src/modal/src/components/ModalFooter.js [hunk 1]
```diff
@@ -27,24 +24,25 @@ const SecondaryButton :any = styled(Button).attrs(() => ({ mode: 'default' }))`
...
-  & OverrideFooterProps;
-
-export default class ModalFooter extends Component<Props> {
+export default class ModalFooter extends Component<ModalFooterProps> {
 
   static propTypes = {
+    isDisabledPrimary: PropTypes.bool,
+    isDisabledSecondary: PropTypes.bool,
+    isPendingPrimary: PropTypes.bool,
+    isPendingSecondary: PropTypes.bool,
```
Comment (solomontang) [inline src/modal/src/components/ModalFooter.js]: This seems like a slippery slope. Perhaps we should err on the side of leniency with the API and be more strict with internal code reviews?

File: src/modal/src/components/Modal.js [hunk 2]
```diff
@@ -15,9 +15,13 @@ import ModalHeader from './ModalHeader';
...
 import { ESC_KEY_CODE } from '../../../utils/keycodes';
 
-type Props = {
+import type { ModalFooterProps } from './ModalFooter';
+import type { ModalHeaderProps } from './ModalHeader';
+
+type ModalProps = {
   children :Node;
   isVisible :boolean;
+  modalRef ?:{ current :null | HTMLElement };
```
[the remark this answers is not in the document]
Comment (UnsungHero97) [reply src/modal/src/components/Modal.js]: It's being used in the orgs app.

File: src/modal/src/components/ActionModal.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  if (requestState === RequestStates.FAILURE) {
+    body = requestStateComponents[RequestStates.FAILURE] || (
+      <span>Failure. Please try again.</span>
+    );
+  }
+  else if (requestState === RequestStates.SUCCESS) {
+    body = requestStateComponents[RequestStates.SUCCESS] || (
+      <span>Success!</span>
+    );
+  }
```
[the remark this answers is not in the document]
Comment (UnsungHero97) [reply src/modal/src/components/ActionModal.js]: I feel icky setting default props to JSX elements. Have you done this before?

File: src/modal/src/components/ModalFooter.test.js [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+        <ModalFooter
+            onClickPrimary={nope}
+            onClickSecondary={nope}
+            textPrimary={CONFIRM_TXT}
+            textSecondary="" />
+      );
+      expect(toJson(wrapper)).toMatchSnapshot();
+    });
+
+    test('should match snapshot with secondary button only', () => {
```
[the remark this answers is not in the document]
Comment (UnsungHero97) [reply src/modal/src/components/ModalFooter.test.js]: I like snapshots for generating the actual JSX for different conditions. In this case, the modal can have only the primary button, or only the secondary button, or both at the same time, or each of those options with stretchy buttons.

The snapshots are useful to visually inspect the JSX for each condition. It's also useful to potentially catch changes in dependencies that might only affect rendering and not our internal component logic.

File: src/modal/src/components/ModalFooter.js (hunk 1 above)
Comment (UnsungHero97) [reply src/modal/src/components/ModalFooter.js]: I do agree, this is not ideal. However, `ModalFooter` is an internal component at the moment so this won't affect users.

File: src/modal/src/components/ActionModal.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (solomontang) [reply src/modal/src/components/ActionModal.js]: Now that I think about it, I have not. 

I have achieved a similar outcome by merging default objects with passed in objects.
https://github.com/openlattice/lattice-ui-kit/blob/develop/src/table/src/components/Table.js#L64

File: src/modal/src/components/ModalFooter.js (hunk 1 above)
Comment (solomontang) [reply src/modal/src/components/ModalFooter.js]: How do you envision this changing if we were to expose the ModalFooter? I have plans to work on different variations of the Modal (ie. `danger`, `warning`, etc.) and it seems like exposing the child components is inevitable.

**Pull request merged**